consensus/src/transaction_shuffler/use_case_aware/iterator.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::use_case_aware::{
+    delayed_queue::DelayedQueue,
+    types::{InputIdx, OutputIdx, UseCaseAwareTransaction},
+    Config,
+};
+use std::{collections::VecDeque, fmt::Debug};
+
+#[derive(Debug)]
+pub(super) struct ShuffledTransactionIterator<Txn> {
+    input_queue: VecDeque<Txn>,
+    delayed_queue: DelayedQueue<Txn>,
+    input_idx: InputIdx,
+    output_idx: OutputIdx,
+}
+
+impl<Txn> ShuffledTransactionIterator<Txn>
+where
+    Txn: UseCaseAwareTransaction + Debug,
+{
+    pub(super) fn new(config: Config) -> Self {
+        Self {
+            input_queue: VecDeque::new(),
+            delayed_queue: DelayedQueue::new(config),
+            input_idx: 0,
+            output_idx: 0,
+        }
+    }
+
+    pub(super) fn extended_with(mut self, txns: impl IntoIterator<Item = Txn>) -> Self {
+        self.input_queue.extend(txns);
+        self
+    }
+
+    pub(super) fn select_next_txn(&mut self) -> Option<Txn> {
+        let ret = self.select_next_txn_inner();
+        if ret.is_some() {
+            self.output_idx += 1;
+        }
+        ret
+    }
+
+    pub(super) fn select_next_txn_inner(&mut self) -> Option<Txn> {
+        // println!("\n### {}: selection started.\n", self.output_idx);
+        // println!("Starting with state:");
+        // println!("{:#?}\n", self);
+
+        self.delayed_queue.bump_output_idx(self.output_idx);
```

Comment:
Instead of passing the outout idx, may be cleaner to increment the output index in the delayed_queue itself?

Reply:
K. It felt right(er) to track the input and output indices at the same level. But I'm fine the other way too.

Reply:
Acutally, if tracked inside the queue, the logic in `select_next_txn()` (that increments the output idx only if something is selected ) will be scattered in multiple calls. The logic is there because I have the execution pool in mind where after a None is selected, new txns can be put into the input queue and selection can be resumed. Does it make sense?